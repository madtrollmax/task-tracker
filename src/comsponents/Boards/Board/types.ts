import { ChangeEvent, EventSimple } from '../../../utils/commonTypes';
import { IColumn } from '../../Columns/types';

export interface IUseBoardViewProps {
    isAdd: boolean;
    startAdd: EventSimple;
    onNewColumnTitleChange: ChangeEvent;
    newColumnTitle: string;
    onCancelAddNewColumn: EventSimple;
    addColumn: EventSimple;
    columns: IColumn[];
};

export type UseBoardFunc = () => IUseBoardViewProps;