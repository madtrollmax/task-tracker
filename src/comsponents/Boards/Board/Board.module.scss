@import '/src/utils/styles.scss';
.board {
    @include border;
    @include scrollbar;
    margin-left: 3px;
    @media (max-width:480px) {
        margin-left: 43px;
    }
    margin-right: 15px;
    padding: 6px;
    overflow-x: scroll;
    overflow-y: scroll;
    flex-grow: 1;

    display: flex;
    gap: 6px;

    .column {
        @include border;
        width: 240px;
        height: 40px;
        display: flex;
        align-items: center;
        justify-content: center;
        flex-shrink: 0;
    }
}