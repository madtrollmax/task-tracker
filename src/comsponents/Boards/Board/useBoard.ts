import { useCallback, useEffect, useMemo, useState } from 'react';
import { useSelector } from 'react-redux';
import { useParams } from 'react-router-dom';
import { boardsSelector } from '../reducer';
import { UseBoardFunc } from './types';
import { boardsActions, path } from '../consts';
import { useAppDispatch } from '../../../utils/hooks/useAppDispatch';
import { createAction, getAction } from '../../../utils/actions/actions';
import { useVisibility } from '../../../utils/hooks/useVisibility'
import { IColumn } from '../../Columns/types';
import { path as columnsPath } from '../../Columns/consts';
import shortid from 'shortid';

export const useBoard: UseBoardFunc = () => {
    const state = useSelector(boardsSelector);
    console.log(state, useSelector);
    const {boardId} = useParams();
    const dispatch = useAppDispatch();

    const{isVisible: isAdd, setVisible: startAdd, setHidden: cancelAdd} = useVisibility(false)
    const [newColumnTitle, setNewColumnTitle ] = useState('');

    const onNewColumnTitleChange = useCallback((event: React.ChangeEvent<HTMLInputElement>) => {
        setNewColumnTitle(event.target.value);
    }, []);

    const onCancelAddNewColumn = useCallback(() => {
        cancelAdd();
        setNewColumnTitle('');
    },[cancelAdd, setNewColumnTitle ]);
    
    useEffect(() => {
        if(!newColumnTitle) dispatch(getAction(boardsActions, boardId as string, path));
    },[dispatch, boardId, newColumnTitle]);

    const addColumn = useCallback(() => {
        dispatch(createAction<IColumn>(boardsActions, { id: shortid.generate(), boardId:(boardId as string), title: newColumnTitle }, columnsPath));
        onCancelAddNewColumn();
    }, [dispatch, newColumnTitle, onCancelAddNewColumn, boardId]);

    const columns = useMemo(() => state.item.columns || [], [state]);


    console.log(state.item.columns);
    

    return {
        isAdd,
        startAdd,
        newColumnTitle,
        onNewColumnTitleChange,
        onCancelAddNewColumn,
        addColumn,
        columns
    };
}