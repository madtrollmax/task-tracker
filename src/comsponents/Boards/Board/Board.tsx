import { FC } from "react";
import { useBoard } from "./useBoard";
import styles from './Board.module.scss';
import { Check, XCircle } from 'react-feather'
import { Column } from "../../Columns/Column/Column";

export const Board: FC = () => {
    const {
        isAdd,
        startAdd,
        newColumnTitle,
        onNewColumnTitleChange,
        onCancelAddNewColumn,
        addColumn,
        columns
    } = useBoard();
    return (<div className={styles.board}>
        {columns.map(el => <Column key={el.id} column={el} />)}
        {isAdd && <div className={styles.column}>
            <input value={newColumnTitle} onChange={onNewColumnTitleChange}/>
            <span>
                <Check onClick={addColumn}/>
                <XCircle onClick={onCancelAddNewColumn}/>
            </span>
        </div>}
        <div className={styles.column} onClick={startAdd}>Создать колонку</div>
    </div>);
}