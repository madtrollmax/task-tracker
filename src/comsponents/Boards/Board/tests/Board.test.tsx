import {render} from '@testing-library/react';
import { Board } from '../Board';

const mockUseBoard = jest.fn().mockReturnValue({
    isAdd: true,
    startAdd: () => null,
    newColumnTitle: 'test title',
    onNewColumnTitleChange: () => null,
    onCancelAddNewColumn: () => null,
    addColumn: () => null,
    columns:[{id:'test_id', title: 'test'}],
});
jest.mock('../useBoard', () => {
    const module = jest.requireActual('../useBoard');
    return {
        ...module,
        useBoard: (...p:any) => mockUseBoard(...p)
    }
});

describe('Board component',  () =>{
    it('default render', ()=> {
        const {asFragment } = render(<Board/>);

        expect(asFragment()).toMatchSnapshot();
    })
})