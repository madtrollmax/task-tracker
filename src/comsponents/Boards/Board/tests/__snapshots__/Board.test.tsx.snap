// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board component default render 1`] = `
<DocumentFragment>
  <div
    class="board"
  >
    <div
      class="column"
    >
      test
    </div>
    <div
      class="column"
    >
      <input
        value="test title"
      />
      <span>
        <svg
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polyline
            points="20 6 9 17 4 12"
          />
        </svg>
        <svg
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="12"
            cy="12"
            r="10"
          />
          <line
            x1="15"
            x2="9"
            y1="9"
            y2="15"
          />
          <line
            x1="9"
            x2="15"
            y1="9"
            y2="15"
          />
        </svg>
      </span>
    </div>
    <div
      class="column"
    >
      Создать колонку
    </div>
  </div>
</DocumentFragment>
`;
